sharpy/aero/models/nonliftingbodygrid.py

```diff
@@ -0,0 +1,116 @@
+            nonlifting_body_tstep.zeta[i_surf], nonlifting_body_tstep.zeta_dot[i_surf] = self.get_zeta_and_zeta_dot(i_surf, structure_tstep)
+
+    def get_zeta_and_zeta_dot(self,i_surf, structure_tstep):
+        numb_radial_nodes = self.dimensions[i_surf][0] +1
+        matrix_nodes = np.zeros((3, numb_radial_nodes,
+                                 self.dimensions[i_surf][1]+1))
+        matrix_nodes_dot = matrix_nodes.copy()
+        array_phi_coordinates = np.linspace(0, 2*np.pi, numb_radial_nodes)
+        # cache sin and cos values
+        array_sin_phi = np.sin(array_phi_coordinates)
+        array_cos_phi = np.cos(array_phi_coordinates)
+
+        cga_rotation_matrix = structure_tstep.cga()
+
+        for node_counter, i_global_node in enumerate(self.aero2struct_mapping[i_surf]):
+            if self.data_dict["shape"].decode() == 'specific':
+                a_ellipse = self.data_dict["a_ellipse"][i_global_node]
+                b_ellipse = self.data_dict["b_ellipse"][i_global_node]
+                z_0 =self.data_dict["z_0_ellipse"][i_global_node]
+                if a_ellipse == 0. or b_ellipse == 0.:
+                    radius = 0
+                else:
+                    radius = a_ellipse*b_ellipse/np.sqrt(
+                        (b_ellipse*array_cos_phi)**2
+                        +(a_ellipse*array_sin_phi)**2)
+            else:
+                radius = self.data_dict["radius"][i_global_node]
+                z_0 = 0
+            
+            
+            # get nodes position in B frame
+            matrix_nodes[1, :, node_counter] = radius*array_cos_phi
+            matrix_nodes[2, :, node_counter] = radius*array_sin_phi + z_0
+            
+            # convert position from B to A frame
+            i_elem, i_local_node = self.get_elment_and_local_node_id(i_surf, i_global_node)
+            psi_node = structure_tstep.psi[i_elem, i_local_node,:]
+            psi_dot_node = structure_tstep.psi_dot[i_elem, i_local_node,:]
+            omega_a = algebra.crv_dot2omega(psi_node, psi_dot_node)
+            if not (psi_node == [0, 0, 0]).all():
+                # just perform roation from B to A if psi not 0
+                Cab = algebra.crv2rotation(psi_node)
+                for idx in range(numb_radial_nodes):
+                    matrix_nodes[:, idx, node_counter] = np.dot(Cab, matrix_nodes[:, idx, node_counter])
+            for dim in range(3):
+                matrix_nodes[dim, :, node_counter] += structure_tstep.pos[i_global_node,dim]
+
+                # get zeta dot in A frame (velocity due to pos_dot)
+                matrix_nodes_dot[dim, :, node_counter] += structure_tstep.pos_dot[i_global_node, dim]
+
+            for idx in range(numb_radial_nodes):
+                # get zeta dot in A frame (velocity due to psi_dot))
+                matrix_nodes_dot[:, idx, node_counter] += (np.dot(algebra.skew(omega_a), matrix_nodes[:, idx, node_counter]))
```

Comment:
I guess this bit is quite related to the original generate_zeta_timestep_info in aerogrid.py - in there I realised the rigid body velocity (linear and angular) for the most general expression for the strip velocities were missing (in eq 54 in screenshot attached). 

Specifically here, it seems that in line 92, matrix_nodes used was already inclusive of the rigid body position -> effectively it's the strip coordinates in G frame already (as processed in line 85) - but then we premultiply with another Cga afterwards in line 96 - possible typo?


<img width="637" alt="Screenshot 2023-09-06 at 10 54 53 AM" src="https://github.com/ImperialCollegeLondon/sharpy/assets/60852830/52fa751c-4c65-4591-8508-d67001243110">

Reply:
Hey Kelvin,
let us talk Friday more about it. Even before this extension o, I think the rigid body velocities are added to zeta dot (calculated in *grid.py) within the UVLM.

Reply:
Just to document the discussion we had here: The strip coordinates are only corrected by the beam coordinates expressed in the A-frame but not G-frame. So, the projection into the G-frame is valid.